File: SearchForARange.cpp
class Solution {
public:
	vector<int> searchRange(int A[], int n, int target) {

		int low = 0;
		int height = n;
		vector<int> res(2,-1);
		int mid;

		while (low < height) {
			mid = (height-low)/2 + low;
			if (A[mid] >= target) {
				height = mid;
			}
			else {
				low = mid+1;
			}
		}
		res[0] = height;

		low = 0;height = n;
		while (low < height) {
			mid = (height-low)/2 + low;
			if (A[mid] > target) {
				height = mid;
			}
			else {
				low = mid+1;
			}
		}

		res[1] = low-1;

		if(res[0] > res[1]) {
			return vector<int>(2,-1);
		}

		if (res[0] == res[1] && A[res[0]] != target) {
			return vector<int>(2,-1);
		}

		return res;


	}
};

File: jiuzhang/search-for-a-range/search-for-a-range.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""

class Solution:
    """
    @param A : a list of integers
    @param target : an integer to be searched
    @return : a list of length 2, [index1, index2]
    """
    def searchRange(self, A, target):
        if len(A) == 0:
            return [-1, -1]
        
        start, end = 0, len(A) - 1
        while start + 1 < end:
            mid = (start + end) / 2
            if A[mid] < target:
                start = mid
            else:
                end = mid
        
        if A[start] == target:
            leftBound = start
        elif A[end] == target:
            leftBound = end
        else:
            return [-1, -1]
        
        start, end = leftBound, len(A) - 1
        while start + 1 < end:
            mid = (start + end) / 2
            if A[mid] <= target:
                start = mid
            else:
                end = mid
        if A[end] == target:
            rightBound = end
        else:
            rightBound = start
        return [leftBound, rightBound]

